File: fileManager/send.js
const io =  require('socket.io-client');
const socket = io("http://131.254.101.229:3000");
const fs = require('fs');
const directory = './data/';

const intervention_elements = ['free', 'intervention_address', 'groundbase_location', 'drone_arrival_time',
                         'intervention_description', 'first_drone_deployed_time', 
                         'end_of_intervention_time', 'intervention_limits', 'intervention_name', 
                         'quality', 'intervention_start_time', 'type_of_intervention'];

const intervention_mandatory = ['end_of_intervention_time', 'intervention_name', 'intervention_start_time'];

const mission_elements = ['free', 'was_changed', 'comments', 'mission_end_time', 'was_finished', 
                    'moving_duration', 'payload_weight', 'pilot_name', 'quality', 
                    'scenario', 'mission_start_time', 'type_of_mission'];

const mission_mandatory = ['mission_start_time', 'mission_end_time', 'scenario'];

const date_regex = "(19[0-9][0-9]|20[0-9][0-9])-(0[1-9]|1[0-2])-(0[1-9]|1[0-9]|2[0-9]|3[0-1])T(0[1-9]|1[0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9])";
const scenario_regex = 'S[1-4]';

const picture_extension = ['.jpg', '.png', 'jpeg'];
const video_extension = ['.mov', '.avi', '.mp4'];

var valid_intervention = true;
var valid_mission = true;

socket.on("msg", (message) => {
	console.log(message);
})

fs.readdir(directory, function(err, items) {
	for (var i=0; i<items.length; i++) {
		// check intervention json
		if(items[i] === 'intervention.json'){
			fs.readFile(directory+items[i], 'utf8', function (err, data) {
			  if (err) throw err;
			  var intervention = JSON.parse(data);
			  // intervention contains all the mandatory elements
			  intervention_mandatory.forEach(function(value){
				  if(typeof intervention[value] === 'undefined'){
					  console.log(value +' is undefined or empty for intervention.json');
					  valid_intervention = false;
				  }
			  })
			  // intervention contains only intervention elements
			  for(var key in intervention){
				  if(intervention_elements.indexOf(key) === -1){
					  console.log(key+ ' is not a valid field in intervention.json');
					  valid_intervention = false;
				  }
			  }
			  // intervention dates are real dates
			  if(intervention['end_of_intervention_time'].match(date_regex)===null || intervention['intervention_start_time'].match(date_regex)===null){
				  console.log('Wrong date format for intervention');
				  valid_intervention = false;
			  }
			  console.log(intervention);
			});
		}
		// check mission json
		if(items[i] === 'mission.json'){
			fs.readFile(directory+items[i], 'utf8', function (err, data) {
			  if (err) throw err;
			  var mission = JSON.parse(data);
			  mission_mandatory.forEach(function(value){
				  if(typeof mission[value] === 'undefined'){
					  console.log(value +' is undefined or empty for mission.json');
					  valid_mission = false;
				  }
			  })
			  for(var key in mission){
				  if(mission_elements.indexOf(key) === -1){
					  console.log(key+ ' is not a valid field in mission.json');
					  valid_mission = false;
				  }
			  }
			  if(mission['mission_end_time'].match(date_regex)===null || mission['mission_start_time'].match(date_regex)===null || mission['scenario'].match(scenario_regex)===null){
				  console.log('Wrong date or scenario format for mission');
				  valid_mission = false;
			  }
			  console.log(mission);
			});
		}
	}
	if(valid_intervention && valid_mission){
		fs.readFile(directory+'intervention.json', function(err,data){
			if(err){console.log(err)}
			socket.emit('sendFile', data, 'intervention.json');
		})
		fs.readFile(directory+'mission.json', function(err,data){
			if(err){console.log(err)}
			socket.emit('sendFile', data, 'mission.json');
		})
		for (var i=0; i<items.length; i++) {
			let name = items[i];
			var extension = name.substr(name.length-4);
			// if the file is sent, we add a prefix 'sent_' to the file name
			// if the file in the directory has already a 'sent_' prefix, we don't send it to the server
			if(name.substr(0,5)!=='sent_' && extension !== 'json'){
				fs.readFile(directory+name, function(err,data){
					if(err){console.log(err)}
					if(picture_extension.indexOf(extension) >= 0 || video_extension.indexOf(extension) >= 0){
						socket.emit('sendFile', data, name);				
					}
					else{
						console.log(name + " is not in one of the allowed format, we did not send it to the server");
					}
				});
				if(extension !== "json" && name !== 'image1.jpg' && name !== 'capture.avi'){
					fs.rename(directory+name, directory+'sent_'+name, function (err) {
						console.log('Rename file error ', err); 
					});
				}
			}
			else {
				console.log(name+'was not sent ! Please remove the sent_ prefix of your file if you want to send it');
			}
		}
	}
	else{
		console.log("Files not sent, check the error messages");
	}
});

File: python/send.py
# -*- coding: utf-8 -*-
"""
Created on Wed Aug 28 11:45:27 2019

@author: llesoil
"""

import socket
import os
import json
import re

host = "131.254.101.229" # host ip
port = 9000 # default port
path = "C:/Users/llesoil/Documents/drone/api/data" # data directory

intervention_elements = ['free', 'intervention_address', 'groundbase_location', 'drone_arrival_time',
                         'intervention_description', 'first_drone_deployed_time', 
                         'end_of_intervention_time', 'intervention_limits', 'intervention_name', 
                         'quality', 'intervention_start_time', 'type_of_intervention']

intervention_mandatory = ['end_of_intervention_time', 'intervention_name', 'intervention_start_time']

mission_elements = ['free', 'was_changed', 'comments', 'mission_end_time', 'was_finished', 
                    'moving_duration', 'payload_weight', 'pilot_name', 'quality', 
                    'scenario', 'mission_start_time', 'type_of_mission']

mission_mandatory = ['mission_start_time', 'mission_end_time', 'scenario']

date_regex = "^(19[0-9][0-9]|20[0-9][0-9])-(0[1-9]|1[0-2])-(0[1-9]|1[0-9]|2[0-9]|3[0-1])T(0[1-9]|1[0-9]|2[0-3]):([0-5][0-9]):([0-5][0-9]),([0-9][0-9][0-9])([+]|[-])[0-2][0-9]:[0-5][0-9]"

picture_extension = ['.jpg', '.png', 'jpeg']
video_extension = ['.mov', '.avi', '.mp4']

directory = os.listdir(path)

def send_file_to_server(file):
    size = len(file)
    if size > 5:
        if file[0:5] !="sent_":
            size = bin(size)[2:].zfill(16) # encode filename size as 16 bit binary
            s.send(size.encode('utf-8'))
            s.send(file.encode('utf-8'))
            filename = os.path.join(path, file)
            filesize = os.path.getsize(filename)
            filesize = bin(filesize)[2:].zfill(32) # encode filesize as 32 bit binary
            s.send(filesize.encode('utf-8'))
            file_to_send = open(filename, 'rb')
            l = file_to_send.read()
            s.sendall(l)
            file_to_send.close()
            print(file+' sent!')
            #if file[len(file)-4:len(file)] != "json":
            #    os.rename(filename,os.path.join(path,"sent_"+file))

def is_valid_date(date):
    return re.search(date_regex, date) != None
    
def is_valid_intervention():
    is_valid = True
    f = open(os.path.join(path, "intervention.json"), 'r')
    inter = json.loads(f.read())
    for el_inter in inter:
        if el_inter not in intervention_elements:
            is_valid = False
    for mandatory_inter in intervention_mandatory:
        if mandatory_inter not in inter:
            is_valid = False
    if not isinstance(inter['intervention_name'], str) or inter['intervention_name'] == "":
        is_valid = False
    if not is_valid_date(inter['intervention_start_time']) or not is_valid_date(inter['end_of_intervention_time']):
        is_valid = False
    if inter["groundbase_location"]['latitude'] == None or inter["groundbase_location"]['longitude'] == None:
        is_valid = False
    return is_valid

def is_valid_mission():
    is_valid = True
    f = open(os.path.join(path, "mission.json"), 'r')
    mission = json.loads(f.read())
    for el_mission in mission:
        if el_mission not in mission_elements:
            is_valid = False
    for mandatory_mission in mission_mandatory:
        if mandatory_mission not in mission:
            is_valid = False
    if re.search("^S[1-4]$", mission['scenario'])==None:
        is_valid = False
    if not is_valid_date(mission['mission_start_time']) or not is_valid_date(mission['mission_end_time']):
        is_valid = False
    return is_valid

if __name__ == "__main__":
    
    s = socket.socket()
    s.connect((host, port))
    
    if "intervention.json" and "mission.json" in directory:
        if is_valid_intervention() and is_valid_mission():
            send_file_to_server("intervention.json")
            send_file_to_server("mission.json")
            
            for file in directory:
                file_extension = file[len(file)-4:len(file)]
                if file_extension in picture_extension or file_extension in video_extension:
                    send_file_to_server(file)
    
    s.close()